stan/math/opencl/ref_type.hpp

```diff
@@ -0,0 +1,71 @@
+#ifndef STAN_MATH_OPENCL_REF_TYPE_HPP
+#define STAN_MATH_OPENCL_REF_TYPE_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/opencl/plain_type.hpp>
+#include <stan/math/prim/meta/ref_type.hpp>
+#include <type_traits>
+
+namespace stan {
+
+namespace internal {
+
+template <typename T>
+struct is_trivial_kg_expression : public std::false_type {};
+
+template <>
+struct is_trivial_kg_expression<int> : public std::true_type {};
+template <>
+struct is_trivial_kg_expression<double> : public std::true_type {};
+template <typename T>
+struct is_trivial_kg_expression<math::matrix_cl<T>> : public std::true_type {};
+template <typename T>
+struct is_trivial_kg_expression<math::load_<T>> : public std::true_type {};
+template <typename T>
+struct is_trivial_kg_expression<math::scalar_<T>> : public std::true_type {};
+template <typename T>
+struct is_trivial_kg_expression<math::constant_<T>> : public std::true_type {};
+template <>
+struct is_trivial_kg_expression<math::row_index> : public std::true_type {};
+template <>
+struct is_trivial_kg_expression<math::col_index> : public std::true_type {};
+template <typename T>
+struct is_trivial_kg_expression<math::calc_if_<false, T>>
+    : public std::true_type {};
+
+template <typename T>
+struct is_trivial_kg_expression<math::as_column_vector_or_scalar_<T>>
+    : public is_trivial_kg_expression<std::decay_t<T>> {};
+template <typename T>
+struct is_trivial_kg_expression<math::block_<T>>
+    : public is_trivial_kg_expression<std::decay_t<T>> {};
+template <typename T, bool Colwise, bool Rowwise>
+struct is_trivial_kg_expression<math::broadcast_<T, Colwise, Rowwise>>
+    : public is_trivial_kg_expression<std::decay_t<T>> {};
+template <typename T>
+struct is_trivial_kg_expression<math::calc_if_<true, T>>
+    : public is_trivial_kg_expression<std::decay_t<T>> {};
```

Comment:
[optional] Only suggestion I here is that for each one of these just do a double check test:

```
EXPECT_TRUE(std::is_same<calc_if_<false, T1>, the_other_expected_type>::value);
EXPECT_TRUE(std::is_same<calc_if_<true, T2>, the_expected_type>::value);
```

So don't worry about the reference behavior or checking values (current tests cover that) -- just check the type calculations.

I guess these things are trivially true here in the code but if something broke in the future it would be easier to figure it out with an explicit test than reverse engineering it from wherever it does break.

Reply:
Half of these definitions are recursive - the value depends on child expression. Also types listed here are just ones that can be trivial. There are many other expressions that can not be trivial. So I think writing tests for ~100 different cases is not worth it, given how simple the implementation is.